tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1186,7 +1213,9 @@ unsigned CGMSHLSLRuntime::AddTypeAnnotation(QualType Ty,
     DxilPayloadAnnotation *payloadAnnotation = nullptr;
     if (ValidatePayloadDecl(RT->getDecl(), *m_pHLModule->GetShaderModel(), CGM.getDiags(), CGM.getCodeGenOpts()))
       payloadAnnotation = dxilTypeSys.AddPayloadAnnotation(ST);
-    return ConstructStructAnnotation(annotation, payloadAnnotation, RD, dxilTypeSys);
+    unsigned size = ConstructStructAnnotation(annotation, payloadAnnotation, RD, dxilTypeSys);
+    // Resources don't count towards cbuffer size.
+    return IsHLSLResourceType(Ty) ? 0 : size;
```

Comment:
For structured buffer case that still goes down this path, return 0 so the resource type itself doesn't count as having a size in the constant buffer.

Reply:
This is a tricky bit.  This case handles StructuredBuffer that goes down this path.  It *does* need an annotation constructed, but then it should return zero for size.  Other resources won't go down this path due to the `if (!IsHLSLStructuredBufferType(Ty) && IsHLSLResourceType(Ty))` above.